src/mds/Server.cc

```diff
@@ -2079,6 +2079,39 @@ void Server::handle_slave_auth_pin_ack(MDRequestRef& mdr, MMDSSlaveRequest *ack)
 // HELPERS
 
 
+/**
+ * check whether we are permitted to complete a request
+ *
+ * Check whether we have permission to perform the operation specified
+ * by mask on the given inode, based on the capability in the mdr's
+ * session.
+ */
+bool Server::check_access(MDRequestRef& mdr, CInode *in, unsigned mask)
+{
+  Session *s = mdr->session;
+
+  // FIXME: populate these with real values
+  string path;
+  int uid = mdr->client_request->get_caller_uid();
+
+  // simple read/write check
+  for (vector<MDSCapGrant>::iterator p = s->auth_caps.grants.begin();
+       p != s->auth_caps.grants.end();
+       ++p) {
+    if (p->match.match(path, uid)) {
+      if (p->spec.allows(mask & MAY_READ,
+			 mask & (MAY_WRITE|MAY_EXECUTE))) {
+	return true;
```

Comment:
On Tue, 2 Jun 2015, Gregory Farnum wrote:
> In src/mds/Server.cc:
> 
> > +bool Server::check_access(MDRequestRef& mdr, CInode *in, unsigned mask)
> > +{
> > +  Session *s = mdr->session;
> > +
> > +  // FIXME: populate these with real values
> > +  string path;
> > +  int uid = mdr->client_request->get_caller_uid();
> > +
> > +  // simple read/write check
> > +  for (vector<MDSCapGrant>::iterator p = s->auth_caps.grants.begin();
> > +       p != s->auth_caps.grants.end();
> > +       ++p) {
> > +    if (p->match.match(path, uid)) {
> > +      if (p->spec.allows(mask & MAY_READ,
> > +			 mask & (MAY_WRITE|MAY_EXECUTE))) {
> > +	return true;
> 
> Why does this need direct access instead of using the check functions we already have? If it does, we need to expand that interface rather than
> reaching through it.

Because for each individual grant, we (will) need to then evaluate unix 
permissions based on the uid and gid list named by that grant.  It'll be 
different for each one.  We can't use a wrapper since it'll return on 
the first match which won't evaluate the permissions.

I suppose we could pass in a callback, but this is so much simpler.